nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py

```diff
@@ -0,0 +1,414 @@
+            kl_div = torch.cat([item["kl"] for item in losses_reduced_per_micro_batch])
+            kl_mean = kl_div.mean()
+
+            rewards_all = torch.cat((rewards_chosen, rewards_rejected))
+            rewards_chosen_mean = rewards_chosen.mean()
+            rewards_rejected_mean = rewards_rejected.mean()
+            rewards_all_mean = rewards_all.mean()
+            rewards_all_std = rewards_all.std()
+            rewards_margins = rewards_chosen_mean.nan_to_num(0) - rewards_rejected_mean.nan_to_num(0)
+
+            # average loss across micro batches
+            loss_mean = torch.as_tensor(
+                [loss_reduced["avg"] for loss_reduced in losses_reduced_per_micro_batch],
+                device=torch.cuda.current_device(),
+            ).mean()
+        else:
+
+            loss_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            kl_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            rewards_chosen_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_rejected_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_all_std = torch.tensor(0.0, device=torch.cuda.current_device())
+            rewards_margins = torch.tensor(0.0, device=torch.cuda.current_device())
+
+        # we can only log on one rank if it is rank zero so we broadcast from last rank
+        torch.distributed.broadcast(loss_mean, get_last_rank())
+        torch.distributed.broadcast(kl_mean, get_last_rank())
+
+        torch.distributed.broadcast(rewards_chosen_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_rejected_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_mean, get_last_rank())
+        torch.distributed.broadcast(rewards_all_std, get_last_rank())
+        torch.distributed.broadcast(rewards_margins, get_last_rank())
+
+        metrics = {
+            "loss": loss_mean,
+            "kl": kl_mean,
+            "rewards_chosen_mean": rewards_chosen_mean,
+            "rewards_rejected_mean": rewards_rejected_mean,
+            "rewards_all_mean": rewards_all_mean,
+            "rewards_all_std": rewards_all_std,
+            "rewards_margin": rewards_margins,
+        }
+
+        # move to CPU
+        metrics = {k: v.item() for k, v in metrics.items()}
+
+        return loss_mean.item(), metrics
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self, zero_grad=False)
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def prepare_for_validation_step(self):
+        prepare_for_validation_step(self)
+
+    def finish_validation_step(self):
+        finish_validation_step(self)
+
+    @torch.no_grad()
+    def get_logprob_batch(self, batch):
+        seq_length = batch["samples"].shape[1]
+
```

Comment:
Add: `num_microbatches = divide(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size)`

Reply:
In KTO as there is only one response per prompt over which we compute the loss. So `self.cfg.kto.log_prob_forward_micro_batch_size=1` anyways so it wouldn't matter. But changing it for consistency with DPO code

Reply:
Ah yes, that's a fair point, in that case, no need to change it, but just a comment to remind people it's just a single response per prompt and not like DPO